File: Net/Framework/DesktopExtensions/CSharp/ArcViewOnlyExtension.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;
using ESRI.ArcGIS.esriSystem;
using ESRI.ArcGIS.Framework;
using ESRI.ArcGIS.ADF.CATIDs;

namespace DesktopExtensionsCS
{
    [Guid("1ce76e5e-261c-48d9-a3f5-3ae8f3c3eac7")]
    [ClassInterface(ClassInterfaceType.None)]
    [ProgId("DesktopExtensionsCS.ArcViewOnlyExtension")]
    public class ArcViewOnlyExtension : IExtension, IExtensionConfig
    {
        #region COM Registration Function(s)
        [ComRegisterFunction()]
        [ComVisible(false)]
        static void RegisterFunction(Type registerType)
        {
            // Required for ArcGIS Component Category Registrar support
            ArcGISCategoryRegistration(registerType);

            //
            // TODO: Add any COM registration code here
            //
        }

        [ComUnregisterFunction()]
        [ComVisible(false)]
        static void UnregisterFunction(Type registerType)
        {
            // Required for ArcGIS Component Category Registrar support
            ArcGISCategoryUnregistration(registerType);

            //
            // TODO: Add any COM unregistration code here
            //
        }

        #region ArcGIS Component Category Registrar generated code
        /// <summary>
        /// Required method for ArcGIS Component Category registration -
        /// Do not modify the contents of this method with the code editor.
        /// </summary>
        private static void ArcGISCategoryRegistration(Type registerType)
        {
            string regKey = string.Format("HKEY_CLASSES_ROOT\\CLSID\\{{{0}}}", registerType.GUID);
            MxExtension.Register(regKey);
            SxExtensions.Register(regKey);
            GxExtensions.Register(regKey);
            GMxExtensions.Register(regKey);
        }
        /// <summary>
        /// Required method for ArcGIS Component Category unregistration -
        /// Do not modify the contents of this method with the code editor.
        /// </summary>
        private static void ArcGISCategoryUnregistration(Type registerType)
        {
            string regKey = string.Format("HKEY_CLASSES_ROOT\\CLSID\\{{{0}}}", registerType.GUID);
            MxExtension.Unregister(regKey);
            SxExtensions.Unregister(regKey);
            GxExtensions.Unregister(regKey);
            GMxExtensions.Unregister(regKey);

        }

        #endregion
        #endregion
        private IApplication m_application;
        private esriExtensionState m_enableState;

        private const esriLicenseProductCode RequiredProductCode = esriLicenseProductCode.esriLicenseProductCodeBasic;
        
        #region IExtension Members

        /// <summary>
        /// Name of extension. Do not exceed 31 characters
        /// </summary>
        public string Name
        {
            get
            {
                return "ArcViewOnlyExtension_CS";
            }
        }

        public void Shutdown()
        {
            m_application = null;
        }

        public void Startup(ref object initializationData)
        {
            m_application = initializationData as IApplication;
            if (m_application == null)
                return;
        }

        #endregion

        #region IExtensionConfig Members

        public string Description
        {
            get
            {
                return "Basic Only Extension (C# Sample)\r\n" +
                    "Copyright © ESRI 2006\r\n\r\n" +
                    "Only available with an Basic product license.";
            }
        }

        /// <summary>
        /// Friendly name shown in the Extension dialog
        /// </summary>
        public string ProductName
        {
            get
            {
                return "Basic Extension (C#)";
            }
        }

        public esriExtensionState State
        {
            get
            {
                return m_enableState;
            }
            set
            {
                if (m_enableState != 0 && value == m_enableState)
                    return;

                //Check if ok to enable or disable extension
                esriExtensionState requestState = value;
                if (requestState == esriExtensionState.esriESEnabled)
                {
                    //Cannot enable if it's already in unavailable state
                    if (m_enableState == esriExtensionState.esriESUnavailable)
                    {
                        throw new COMException("Not running the appropriate product license.");
                    }

                    //Determine if state can be changed
                    esriExtensionState checkState = StateCheck(true);
                    m_enableState = checkState;
                    if (m_enableState == esriExtensionState.esriESUnavailable)
                        throw new COMException("Not running the appropriate product license.");
                }
                else if (requestState == 0 || requestState == esriExtensionState.esriESDisabled)
                {
                    //Determine if state can be changed
                    esriExtensionState checkState = StateCheck(false);
                    if (checkState != m_enableState)
                        m_enableState = checkState;
                }

            }
        }

        #endregion

        /// <summary>
        /// Determine extension state 
        /// </summary>
        /// <param name="requestEnable">true if to enable; false to disable</param>
        private esriExtensionState StateCheck(bool requestEnable)
        {
            //Turn on or off extension directly
            if (requestEnable)
            {
                //Check if the correct product is licensed
                IAoInitialize aoInitTestProduct = new AoInitializeClass();
                esriLicenseProductCode prodCode = aoInitTestProduct.InitializedProduct();
                if (prodCode == RequiredProductCode)
                    return esriExtensionState.esriESEnabled;

                return esriExtensionState.esriESUnavailable;
            }
            else
                return esriExtensionState.esriESDisabled;
        }

    }
}

File: Net/Framework/DesktopExtensions/VBNet/ArcViewOnlyExtension.vb
Imports ESRI.ArcGIS.ADF.CATIDs
Imports ESRI.ArcGIS.esriSystem
Imports System.Runtime.InteropServices
Imports ESRI.ArcGIS.Framework

<ComClass(ArcViewOnlyExtension.ClassId, ArcViewOnlyExtension.InterfaceId, ArcViewOnlyExtension.EventsId), _
 ProgId("DesktopExtensionsVB.ArcViewOnlyExtension")> _
Public Class ArcViewOnlyExtension
    Implements IExtension
    Implements IExtensionConfig

#Region "COM Registration Function(s)"
    <ComRegisterFunction(), ComVisibleAttribute(False)> _
    Public Shared Sub RegisterFunction(ByVal registerType As Type)
        ' Required for ArcGIS Component Category Registrar support
        ArcGISCategoryRegistration(registerType)

        'Add any COM registration code after the ArcGISCategoryRegistration() call

    End Sub

    <ComUnregisterFunction(), ComVisibleAttribute(False)> _
    Public Shared Sub UnregisterFunction(ByVal registerType As Type)
        ' Required for ArcGIS Component Category Registrar support
        ArcGISCategoryUnregistration(registerType)

        'Add any COM unregistration code after the ArcGISCategoryUnregistration() call

    End Sub

#Region "ArcGIS Component Category Registrar generated code"
    ''' <summary>
    ''' Required method for ArcGIS Component Category registration -
    ''' Do not modify the contents of this method with the code editor.
    ''' </summary>
    Private Shared Sub ArcGISCategoryRegistration(ByVal registerType As Type)
        Dim regKey As String = String.Format("HKEY_CLASSES_ROOT\CLSID\{{{0}}}", registerType.GUID)
        MxExtension.Register(regKey)
        GxExtensions.Register(regKey)
        SxExtensions.Register(regKey)
        GMxExtensions.Register(regKey)

    End Sub
    ''' <summary>
    ''' Required method for ArcGIS Component Category unregistration -
    ''' Do not modify the contents of this method with the code editor.
    ''' </summary>
    Private Shared Sub ArcGISCategoryUnregistration(ByVal registerType As Type)
        Dim regKey As String = String.Format("HKEY_CLASSES_ROOT\CLSID\{{{0}}}", registerType.GUID)
        MxExtension.Unregister(regKey)
        GxExtensions.Unregister(regKey)
        SxExtensions.Unregister(regKey)
        GMxExtensions.Unregister(regKey)

    End Sub

#End Region
#End Region

#Region "COM GUIDs"
    ' These  GUIDs provide the COM identity for this class 
    ' and its COM interfaces. If you change them, existing 
    ' clients will no longer be able to access the class.
    Public Const ClassId As String = "9ef36120-9c61-4207-9884-601a19f25d49"
    Public Const InterfaceId As String = "347f8040-acb7-4d92-b371-880ed3964bd8"
    Public Const EventsId As String = "442b453d-64e9-4a2c-ad4e-141973880eb2"
#End Region

    Private m_application As IApplication
    Private m_enableState As esriExtensionState

    Private Const RequiredProductCode As esriLicenseProductCode = esriLicenseProductCode.esriLicenseProductCodeBasic

    ' A creatable COM class must have a Public Sub New() 
    ' with no parameters, otherwise, the class will not be 
    ' registered in the COM registry and cannot be created 
    ' via CreateObject.
    Public Sub New()
        MyBase.New()
    End Sub

    ''' <summary>
    ''' Determine extension state
    ''' </summary>
    Private Function StateCheck(ByVal requestEnable As Boolean) As esriExtensionState
        'Turn on or off extension directly 
        If requestEnable Then
            'Check if the correct product is licensed
            Dim aoInitTestProduct As IAoInitialize = New AoInitializeClass()
            Dim prodCode As esriLicenseProductCode = aoInitTestProduct.InitializedProduct()
            If prodCode = RequiredProductCode Then _
                Return esriExtensionState.esriESEnabled

            Return esriExtensionState.esriESUnavailable
        Else
            Return esriExtensionState.esriESDisabled
        End If
    End Function

#Region "IExtension Members"
    ''' <summary>
    ''' Name of extension. Do not exceed 31 characters
    ''' </summary>
    Public ReadOnly Property Name() As String Implements ESRI.ArcGIS.esriSystem.IExtension.Name
        Get
            Return "ArcViewOnlyExtension_VB"
        End Get
    End Property

    Public Sub Shutdown() Implements ESRI.ArcGIS.esriSystem.IExtension.Shutdown
        m_application = Nothing
    End Sub

    Public Sub Startup(ByRef initializationData As Object) Implements ESRI.ArcGIS.esriSystem.IExtension.Startup
        m_application = CType(initializationData, IApplication)
        If m_application Is Nothing Then Return

    End Sub
#End Region

#Region "IExtensionConfig Members"
    Public ReadOnly Property Description() As String Implements ESRI.ArcGIS.esriSystem.IExtensionConfig.Description
        Get
            Return "Basic Only Extension (VB.Net Sample)\r\n" + _
                "Copyright © ESRI 2006\r\n\r\n" + _
                "Only available with an Basic product license."
        End Get
    End Property

    ''' <summary>
    ''' Friendly name shown in the Extensions dialog
    ''' </summary>
    Public ReadOnly Property ProductName() As String Implements ESRI.ArcGIS.esriSystem.IExtensionConfig.ProductName
        Get
            Return "Basic Extension (VB.Net)"
        End Get
    End Property

    Public Property State() As ESRI.ArcGIS.esriSystem.esriExtensionState Implements ESRI.ArcGIS.esriSystem.IExtensionConfig.State
        Get
            Return m_enableState
        End Get
        Set(ByVal value As ESRI.ArcGIS.esriSystem.esriExtensionState)
            If m_enableState <> 0 And value = m_enableState Then Exit Property

            'Check if ok to enable or disable extension
            Dim requestState As esriExtensionState = value
            If requestState = esriExtensionState.esriESEnabled Then
                'Cannot enable if it's already in unavailable state
                If m_enableState = esriExtensionState.esriESUnavailable Then
                    Throw New COMException("Not running the appropriate product license.")
                End If

                'Determine if state can be changed
                Dim checkState As esriExtensionState = StateCheck(True)
                m_enableState = checkState

                If m_enableState = esriExtensionState.esriESUnavailable Then
                    Throw New COMException("Not running the appropriate product license.")
                End If

            ElseIf requestState = esriExtensionState.esriESDisabled Then
                'Determine if state can be changed
                Dim checkState As esriExtensionState = StateCheck(False)
                If (m_enableState <> checkState) Then m_enableState = checkState
            End If
        End Set
    End Property
#End Region


End Class


